File: tests/r_parity/75_stacked.py
"""StatsPAI stacked-DiD parity (Python side) -- Module 75.

Runs ``sp.stacked_did`` under both control-group conventions; the companion R
script builds the same stacks by hand and fits them with ``fixest::feols``.

There is no CRAN package for Cengiz-Dube-Lindner-Zipperer stacking, so the R
side is a hand-written reference rather than a package call. What it pins is
that StatsPAI's stacking construction (one sub-experiment per cohort, clean
controls over the window, cohort-specific unit and time fixed effects,
k = -1 as reference) reproduces the same design when written independently.

Tolerance: rel < 1e-6 on the event-study coefficients and the post mean.
"""

from __future__ import annotations

import warnings

import numpy as np
import pandas as pd

from _common import ParityRecord, dump_csv, write_results

MODULE = "75_stacked"
WINDOW = (-3, 3)


def _panel() -> pd.DataFrame:
    rng = np.random.default_rng(23)
    rows = []
    for u in range(240):
        g = int(rng.choice([6, 9, 12, 0]))  # 0 = never treated
        fe = rng.normal(0, 0.8)
        for t in range(1, 17):
            on = 1 if (g > 0 and t >= g) else 0
            rows.append(
                {
                    "id": u,
                    "year": t,
                    "first_treat": g,
                    "y": fe + 0.15 * t + 1.3 * on + rng.normal(0, 0.6),
                }
            )
    return pd.DataFrame(rows)


def main() -> None:
    import statspai as sp

    df = _panel()
    dump_csv(df, MODULE)

    rows: list[ParityRecord] = []
    for spec, never_only in (("never", True), ("nyt", False)):
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            fit = sp.stacked_did(
                df,
                y="y",
                group="id",
                time="year",
                first_treat="first_treat",
                window=WINDOW,
                never_treated_only=never_only,
            )
        n = int(fit.model_info["n_stacked_obs"])
        es = fit.model_info["event_study"]
        for _, row in es.iterrows():
            k = int(row["relative_time"])
            if k == -1:  # reference period, not estimated
                continue
            rows.append(
                ParityRecord(
                    module=MODULE,
                    side="py",
                    statistic=f"{spec}_att_rel_{k}",
                    estimate=float(row["att"]),
                    se=float(row["se"]),
                    n=n,
                )
            )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"{spec}_ATT_post",
                estimate=float(fit.estimate),
                se=None,
                n=n,
            )
        )

    write_results(MODULE, "py", rows, extra={"estimator": "stacked_did"})


if __name__ == "__main__":
    main()

File: tests/r_parity/75_stacked.R
# StatsPAI stacked-DiD parity (R side) -- Module 75.
#
# There is no CRAN package for Cengiz-Dube-Lindner-Zipperer stacking, so this
# is a hand-written reference: one sub-experiment per treated cohort with its
# clean controls over the event window, stacked, then TWFE with
# cohort-specific unit and time fixed effects and k = -1 as the reference
# period. Estimated with fixest::feols, clustered on the unit id.
#
# Both control-group conventions are emitted: never-treated only (StatsPAI's
# default) and never + not-yet-treated.
#
# Tolerance: rel < 1e-6 on the event-study coefficients and the post-period
# mean.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(fixest)
  library(data.table)
})

MODULE <- "75_stacked"
W_LO <- -3
W_HI <- 3

df <- as.data.table(read_csv_strict(MODULE))
cohorts <- sort(unique(df[first_treat > 0, first_treat]))
never <- unique(df[first_treat == 0, id])

stack_fit <- function(never_only) {
  frames <- list()
  for (g in cohorts) {
    t_lo <- g + W_LO
    t_hi <- g + W_HI
    coh <- unique(df[first_treat == g, id])
    nyt <- unique(df[first_treat > t_hi & first_treat > 0, id])
    ctrl <- if (never_only) never else union(never, nyt)
    if (!length(ctrl)) next
    sub <- df[id %in% union(coh, ctrl) & year >= t_lo & year <= t_hi]
    if (!nrow(sub)) next
    sub[, `:=`(
      cohort = g,
      rel = year - g,
      treated_unit = as.integer(id %in% coh)
    )]
    frames[[length(frames) + 1]] <- sub
  }
  st <- rbindlist(frames)
  st[, `:=`(uc = paste0(id, "_", cohort), tc = paste0(year, "_", cohort))]
  st[, rel_f := factor(rel, levels = sort(unique(rel)))]
  fit <- feols(y ~ i(rel_f, treated_unit, ref = "-1") | uc + tc,
               data = st, cluster = ~ id)
  list(fit = fit, n = nrow(st))
}

rows <- list()
for (spec in c("never", "nyt")) {
  res <- stack_fit(spec == "never")
  co <- coef(res$fit)
  se <- se(res$fit)
  rels <- as.numeric(sub(".*rel_f::(-?[0-9]+):.*", "\\1", names(co)))
  for (k in seq_along(co)) {
    rows[[length(rows) + 1]] <- parity_row(
      module    = MODULE,
      statistic = sprintf("%s_att_rel_%d", spec, rels[k]),
      estimate  = unname(co[k]),
      se        = unname(se[k]),
      n         = res$n
    )
  }
  rows[[length(rows) + 1]] <- parity_row(
    module    = MODULE,
    statistic = sprintf("%s_ATT_post", spec),
    estimate  = mean(co[rels >= 0]),
    se        = NA_real_,
    n         = res$n
  )
}

write_results(MODULE, rows, extra = list(
  reference = "hand-written stacked DiD via fixest::feols",
  fixest = as.character(utils::packageVersion("fixest"))
))
